### PR TITLE
feat: add node address cache

### PR DESCRIPTION
**Pull request by jacobkaufmann**
### What was wrong?

If we receive a talk request from some unknown node, we only [see](https://github.com/sigp/discv5/blob/1b66fd49a86db1193c73af247f19f99626eaab02/src/service.rs#L60) that node's [ID and socket address](https://github.com/sigp/discv5/blob/1b66fd49a86db1193c73af247f19f99626eaab02/src/node_info.rs#L150). There is no way to look up the ENR for that node given that info. This was not even possible until [this change](https://github.com/sigp/discv5/pull/123) to `sigp/discv5`.

### How was it fixed?

- Define `NodeAddress` to hold the observed contact info for a remote node (ENR and socket address).
- Add a LRU cache in `Discovery` to hold `NodeAddress` values.
- Update cache based on sessions established by the underlying `Discv5` service.
- Add `cached_node_addr` method to `Discovery` to provide read-only access to the cache.

### To-Do

[//]: # (Stay ahead of things, add list items here!)
- [x] Add entry to the [release notes](https://github.com/ethereum/trin/blob/master/newsfragments/README.md) (may forgo for trivial changes)
- [x] Clean up commit history

### CODE REVIEW

**Comment by jacobkaufmann** [comment]
Some things I'm unsure of:

- Given that the cache holds `(Enr, SocketAddr)` entries, should we move away from the ENR-centric naming?
- What should the capacity of the cache be? Should it be a function of the routing table bucket size?

File: trin-core/src/portalnet/types/messages.rs [hunk 1]
```diff
@@ -129,6 +130,7 @@ impl Default for PortalnetConfig {
             data_radius: U256::from(u64::MAX), //TODO better data_radius default?
             no_stun: false,
             enable_metrics: false,
+            enr_cache_capacity: 256,
```
Comment (ogenev) [inline trin-core/src/portalnet/types/messages.rs:133]: I think 256 is a reasonable initial value. I guess we expect only the first 7-8 buckets to include any nodes, so if we take the max capacity of 16 nodes per bucket and 8 buckets x 2 networks = 256.

**Review by carver — APPROVED** [review]
> Given that the cache holds `(Enr, SocketAddr)` entries, should we move away from the ENR-centric naming?

Yeah, it seems reasonable to me to make a `NodeAddress` struct that has those to fields. Then it would be a "node address cache". No strong feelings about whether to include that in this PR. If that sounds good to you then 👍🏻

File: Cargo.toml [hunk 2]
```diff
@@ -11,9 +11,11 @@ repository = "https://github.com/ethereum/trin"
 
 [dependencies]
 discv5 = { git = "https://github.com/sigp/discv5.git", branch = "master" }
+enr = { version = "=0.6.0", features = ["k256", "ed25519"] }
```
[the remark this answers is not in the document]
Comment (jacobkaufmann) [reply Cargo.toml:14]: I was running into build failures due to transitive dependencies of `enr`, which is why I had to lock the version here.

**Comment by jacobkaufmann** [comment]
@carver I went with your suggestion and defined a `NodeAddress` here, changing the naming of the cache around that struct. It does not mirror the struct by the same name in `discv5`, but that struct is private so there should not be a conflict there.

@ogenev and as far as the cache capacity I defined a constant equal to 512 that depends on the Discv5 k-bucket size. I made it so that there is enough capacity for 32 full buckets, which may be shared among multiple subnetworks.

**Pull request merged**